Add overlay service to dispatch and handle requests

jacobkaufmann opened the pull request:
Related to #135. Move message dispatch and handling into `OverlayService`. The `Overlay` communicates with the associated `OverlayService` via a channel.

Motivation: Autonomous routing table management requires a separate task where the task must have a mutable reference to the overlay structure. We create a new overlay structure, `OverlayService`, and communication between this structure and `Overlay` occurs over a channel, which does not require a reference.

njgheorghita commented:
@jacobkaufmann Yeah, it's a bit large but also sometimes that's unavoidable. It also seem like it covers some of the same territory as #163, #164 - so maybe wait until those land (which should be soon) and then either rebase off them, or leave your changes here if you feel like it's a more suitable error structure (which very well might be the case).

njgheorghita reviewed (approved):
Looks great! Nice work, just one question, so it's mentioned that the overlay routing table maintenance / liveness tasks are done autonomously - which I interpret as handled within `discv5` - so my understanding here is that the maintenance / liveness tasks are done over the base discv5 protocol? This seems fine for `PING`, but are `FINDNODES` lookup/bootstrapping messages part of the autonomous management at all? If so, and those are going out over the base discv5 protocol, that would be corrupt the routing table

jacobkaufmann commented on ethportal-peertest/src/main.rs:
```diff
@@ -34,16 +34,37 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         discovery.write_with_warn().await.start().await.unwrap();
 
+        let protocol_rx = discovery
+            .write()
```
Ah yes, good catch, this was a miss by me.

jacobkaufmann commented on trin-history/src/events.rs:
```diff
@@ -29,12 +29,12 @@ impl HistoryEvents {
                 Err(e) => {
                     error!("failed to process portal history event: {}", e);
-                    e.into_bytes()
+                    e.to_string().into_bytes()
```
Yeah, I agree. Initially I changed everything to be `Err(error)` and not `Err(e)`, but I think I switched back to avoid unrelated changes. Looks like I missed some of those. I'll go ahead and update the `Err(e)` to `Err(error)` in parts of the code that this PR touches. I agree that we should standardize a convention in a later PR.

jacobkaufmann commented on ethportal-peertest/src/events.rs:
```diff
@@ -49,14 +46,26 @@ impl PortalnetEvents {
+                        let reply = match self.history_overlay.process_one_request(&request).await {
+                            Ok(r) => Message::Response(r).to_bytes(),
+                            Err(e) => {
+                                error!("failed to process history portal event: {}", e);
```
I agree here. I just kept the pre-existing message, but I can go ahead and update that.

jacobkaufmann commented:
> Looks great! Nice work, just one question, so it's mentioned that the overlay routing table maintenance / liveness tasks are done autonomously - which I interpret as handled within `discv5` - so my understanding here is that the maintenance / liveness tasks are done over the base discv5 protocol? This seems fine for `PING`, but are `FINDNODES` lookup/bootstrapping messages part of the autonomous management at all? If so, and those are going out over the base discv5 protocol, that would be corrupt the routing table

@njgheorghita I think you are referring to the comment above `OverlayService::spawn`? So maybe the remark about "autonomous" processes here is premature, because those processes are not implemented in this PR. If you see #135 in the `overlay/service.rs` file, the main loop performs pings based on the ping queue, (#139 introduces the necessary data structure for the ping queue), and the loop also performs some maintenance tasks on the routing table for previously applied pending entries.

These processes are considered "autonomous" because we do not need to invoke a method on `OverlayService` for them to occur. We just have to spawn the service. Also, these processes do not occur over `discv5` `PING` or `FINDNODES` messages. The ping messages that are sent are overlay `PING` messages, and the node updates are done via overlay `FINDNODES` messages that contain the latest ENR.

I hope this clarifies things. If the autonomous comment is confusing without the implementation of those processes, then I'm happy to remove it. And if anything else deserves some further elaboration, then I can add additional comments.

njgheorghita commented:
@jacobkaufmann Ahh cool, yeah that makes perfect sense, thanks!

ogenev commented on ethportal-peertest/src/events.rs:
```diff
@@ -49,14 +46,26 @@ impl PortalnetEvents {
             match std::str::from_utf8(request.protocol()) {
                 Ok(protocol) => match protocol {
```
FYI this match will not work with the latest ProtocolId update in https://github.com/ethereum/trin/pull/162. I'll create an issue to fix this in a separate PR.

jacobkaufmann commented:
I think I've addressed all the comments. Waiting on #176, which @ogenev created to address the comment above, to land before I merge to ensure there are no conflicts.
